exporter/exporterhelper/common.go

```diff
@@ -157,48 +157,106 @@ func WithCapabilities(capabilities consumer.Capabilities) Option {
+	unmarshaler internal.RequestUnmarshaler, osf obsrepSenderFactory, options ...Option) (*baseExporter, error) {
 
-	var err error
-	be.obsrep, err = newObsExporter(obsreport.ExporterSettings{ExporterID: set.ID, ExporterCreateSettings: set}, globalInstruments)
+	obsrep, err := newObsExporter(obsreport.ExporterSettings{ExporterID: set.ID, ExporterCreateSettings: set}, globalInstruments)
 	if err != nil {
 		return nil, err
 	}
 
-	be.qrSender = newQueuedRetrySender(set.ID, signal, bs.queue, bs.RetrySettings, &timeoutSender{cfg: bs.TimeoutSettings}, set.Logger)
-	be.sender = be.qrSender
-	be.StartFunc = func(ctx context.Context, host component.Host) error {
-		// First start the wrapped exporter.
-		if err := bs.StartFunc.Start(ctx, host); err != nil {
-			return err
-		}
+	be := &baseExporter{
+		requestExporter: requestExporter,
+		marshaler:       marshaler,
+		unmarshaler:     unmarshaler,
+		signal:          signal,
 
-		// If no error then start the queuedRetrySender.
-		return be.qrSender.start(ctx, host, set)
+		queueSender:   &baseRequestSender{},
+		obsrepSender:  osf(obsrep),
+		retrySender:   &baseRequestSender{},
+		timeoutSender: &timeoutSender{cfg: NewDefaultTimeoutSettings()},
+
+		set:           set,
+		obsrep:        obsrep,
+		sampledLogger: createSampledLogger(set.Logger),
 	}
-	be.ShutdownFunc = func(ctx context.Context) error {
-		// First shutdown the queued retry sender
-		be.qrSender.shutdown()
-		// Last shutdown the wrapped exporter itself.
-		return bs.ShutdownFunc.Shutdown(ctx)
+
+	for _, op := range options {
+		op(be)
 	}
+	be.connectSenders()
+
 	return be, nil
 }
 
-// wrapConsumerSender wraps the consumer sender (the sender that uses retries and timeout) with the given wrapper.
-// This can be used to wrap with observability (create spans, record metrics) the consumer sender.
-func (be *baseExporter) wrapConsumerSender(f func(consumer requestSender) requestSender) {
-	be.qrSender.consumerSender = f(be.qrSender.consumerSender)
+// send sends the request using the first sender in the chain.
+func (be *baseExporter) send(req internal.Request) error {
+	return be.queueSender.send(req)
+}
+
+// connectSenders connects the senders in the predefined order.
+func (be *baseExporter) connectSenders() {
+	be.queueSender.setNextSender(be.obsrepSender)
+	be.obsrepSender.setNextSender(be.retrySender)
+	be.retrySender.setNextSender(be.timeoutSender)
+}
+
+func (be *baseExporter) Start(ctx context.Context, host component.Host) error {
+	// First start the wrapped exporter.
+	if err := be.StartFunc.Start(ctx, host); err != nil {
+		return err
+	}
+
+	// If no error then start the queueSender.
+	return be.queueSender.start(ctx, host, be.set)
+}
+
+func (be *baseExporter) Shutdown(ctx context.Context) error {
+	// First shutdown the retry sender
+	be.retrySender.shutdown()
+
+	// Then shutdown the queue sender
+	be.queueSender.shutdown()
```

Comment:
Random senders that you start/stop. If the interface is like that, you should call start on all. Also should it call start on the next sender itself and same for stop?

Reply:
The shutdown sequence is a bit random for a reason. Ideally, it could've been the same sequence as in the pipelines, the same as the data flow: queueSender -> retrySender -> user's exporter. But we need to stop the `retrySender` first so it can push an unfinished request back to the queue.

The start sequence should be applied in reverse order.